libr/asm/p/asm_picbaseline.c

```diff
@@ -0,0 +1,328 @@
+		return PIC_BASELINE_OPCODE_XORWF;
+	case 0b010000:
+	case 0b010001:
+	case 0b010010:
+	case 0b010011:
+		return PIC_BASELINE_OPCODE_BCF;
+	case 0b010100:
+	case 0b010101:
+	case 0b010110:
+	case 0b010111:
+		return PIC_BASELINE_OPCODE_BSF;
+	case 0b011000:
+	case 0b011001:
+	case 0b011010:
+	case 0b011011:
+		return PIC_BASELINE_OPCODE_BTFSC;
+	case 0b011100:
+	case 0b011101:
+	case 0b011110:
+	case 0b011111:
+		return PIC_BASELINE_OPCODE_BTFSS;
+	case 0b111000:
+	case 0b111001:
+	case 0b111010:
+	case 0b111011:
+		return PIC_BASELINE_OPCODE_ANDLW;
+	case 0b100100:
+	case 0b100101:
+	case 0b100110:
+	case 0b100111:
+		return PIC_BASELINE_OPCODE_CALL;
+	case 0b101000:
+	case 0b101001:
+	case 0b101010:
+	case 0b101011:
+	case 0b101100:
+	case 0b101101:
+	case 0b101110:
+	case 0b101111:
+		return PIC_BASELINE_OPCODE_GOTO;
+	case 0b110100:
+	case 0b110101:
+	case 0b110110:
+	case 0b110111:
+		return PIC_BASELINE_OPCODE_IORLW;
+	case 0b110000:
+	case 0b110001:
+	case 0b110010:
+	case 0b110011:
+		return PIC_BASELINE_OPCODE_MOVLW;
+	case 0b100000:
+	case 0b100001:
+	case 0b100010:
+	case 0b100011:
+		return PIC_BASELINE_OPCODE_RETLW;
+	case 0b111100:
+	case 0b111101:
+	case 0b111110:
+	case 0b111111:
+		return PIC_BASELINE_OPCODE_XORLW;
+	default:
+		return PIC_BASELINE_OPCODE_INVALID;
+	}
+}
+
+static int asm_picbaseline_disassemble(RAsm *a, RAsmOp *op, const ut8 *b, int l) {
+#define EMIT_INVALID { \
+	op->size = 1; \
+	strncpy (op->buf_asm, "invalid", R_ASM_BUFSIZE); \
+	return 1; \
+}
+	if (!b || l<2) {
+		EMIT_INVALID
+	}
+
+	ut16 instr = r_read_le16 (b);
+	PicBaselineOpcode opcode = picbaseline_get_opcode (instr);
+	if (opcode == PIC_BASELINE_OPCODE_INVALID) {
+		EMIT_INVALID
+	}
+
+	const PicBaselineOpInfo *op_info = &picbaseline_op_info[opcode];
+	op->size = 2;
+
+	switch (op_info->args) {
+	case PIC_BASELINE_OP_ARGS_NONE:
+		strncpy (op->buf_asm, op_info->mnemonic, R_ASM_BUFSIZE);
+		break;
+	case PIC_BASELINE_OP_ARGS_2F:
+		snprintf (op->buf_asm, R_ASM_BUFSIZE + 1, "%s 0x%x",
+				  op_info->mnemonic,
+				  instr & 0b11);
+		break;
+	case PIC_BASELINE_OP_ARGS_3F:
+		snprintf (op->buf_asm, R_ASM_BUFSIZE + 1, "%s 0x%x",
+				  op_info->mnemonic,
+				  instr & 0b111);
+		break;
+	case PIC_BASELINE_OP_ARGS_3K:
+		snprintf (op->buf_asm, R_ASM_BUFSIZE + 1, "%s 0x%x",
+				  op_info->mnemonic,
+				  instr & 0b111);
+		break;
+	case PIC_BASELINE_OP_ARGS_1D_5F:
+		snprintf (op->buf_asm, R_ASM_BUFSIZE + 1, "%s 0x%x, %c",
+				  op_info->mnemonic,
+				  instr & 0b11111,
+				  (instr & (1 << 5)) >> 5 ? 'f' : 'w');
+		break;
+	case PIC_BASELINE_OP_ARGS_5F:
+		snprintf (op->buf_asm, R_ASM_BUFSIZE + 1, "%s 0x%x",
+				  op_info->mnemonic,
+				  instr & 0b11111);
+		break;
+	case PIC_BASELINE_OP_ARGS_3B_5F:
+		snprintf (op->buf_asm, R_ASM_BUFSIZE + 1, "%s 0x%x, 0x%x",
+				  op_info->mnemonic,
+				  instr & 0b11111,
+				  (instr & (0b111 << 5)) >> 5);
+		break;
+	case PIC_BASELINE_OP_ARGS_8K:
+		snprintf (op->buf_asm, R_ASM_BUFSIZE + 1, "%s 0x%x",
```

Comment:
this can result on non-null terminated string

Reply:
No, snprintf takes the number of bytes, not characters. But it would make sense to replace it with sizeof(...).